app/internal/page/multi_read_closer.go

```diff
@@ -0,0 +1,99 @@
+// Modified from io/multi.go which carries the following licence:
+//
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+package page
+
+import (
+	"errors"
+	"io"
+)
+
+type eofReadCloser struct{}
+
+func (eofReadCloser) Read([]byte) (int, error) {
+	return 0, io.EOF
+}
+
+func (eofReadCloser) Close() error {
+	return nil
+}
+
+type multiReadCloser struct {
+	readClosers []io.ReadCloser
+}
+
+func (mr *multiReadCloser) Read(p []byte) (n int, err error) {
+	for len(mr.readClosers) > 0 {
+		// Optimization to flatten nested multiReaders (Issue 13558).
+		if len(mr.readClosers) == 1 {
+			if r, ok := mr.readClosers[0].(*multiReadCloser); ok {
+				mr.readClosers = r.readClosers
+				continue
+			}
+		}
+		n, err = mr.readClosers[0].Read(p)
+		if err == io.EOF {
+			// Use eofReader instead of nil to avoid nil panic
+			// after performing flatten (Issue 18232).
```

Comment:
Is this why we've bought in a modified version?

Reply:
oh no, this could probably be cleaned up. The modification is to take something that handles `io.Reader`s (https://pkg.go.dev/io#MultiReader) to do `io.ReadCloser`s instead